src/locationSource/LocationSource.js

```diff
@@ -0,0 +1,57 @@
+import GpsApiLocationSource from "./GpsApiLocationSource";
+import IpBasedLocationSource from "./IpBasedLocationSource";
+
+export default class LocationSource {
+    async checkLocation() {
+        let location = await this._getGpsLocation();
+
+        if (!location) location = await this._getIPLocation();
+
+        if (location) return location;
+        else throw "Unable to determine the location"
+    }
+
+    async _getGpsLocation() {
+        let location;
+
+        try {
+            location = await this._getGpsLocationWithTimeout();
+        } catch (e) {
+            console.error(e);
+        }
+
+        return location;
+    }
+
+    async _getGpsLocationWithTimeout() {
+        let gps = new GpsApiLocationSource();
+
+        return await Promise.race([
+            gps.checkLocation(),
+            this._timeout(4000)
+        ]);
+    }
+
+    async _getIPLocation() {
+        let location;
+
+        try {
+            location = await this._getIPLocationWithTimeout();
+        } catch (e) {
+            console.error(e);
+        }
+
+        return location;
+    }
+
+    async _getIPLocationWithTimeout() {
+        return await Promise.race([
```

Comment:
Piszę tutaj, ale to jest ogólny problem, który może mieć zastosowanie przy innych użyciach Promise.race() (ale nie wczytywałem się na tyle, żeby mieć pewność ;) ). To nie będzie działało tak, jak (o ile dobrze rozumiem intencję) się spodziewasz. Promise.race() resolvuje na pierwszym Promise które zresolvuje, **ale również rejectuje na pierwszym Promise, które zrejectuje**.
Czyli, w tym konkretnym przypadku, jeśli pierwsza odpowiedź, jaką dostaniemy, poskutkuje rejectem (bo np. jakieś bzdury przyjdą w danych, albo nam odrzuci, albo adblock w ogóle zablokuje wysłanie zapytania) to nawet, jeśli drugi serwis odpowie poprawnymi danymi, skutku to żadnego mieć nie będzie - bo całe Promise.race już zrejectowało.

Reply:
To że rejectuje na pierwszym Promise też jest nam na rękę, bo dostalibyśmy odmowę przed końcem timeoutu. Myślisz że w takim wypadku wystarczyłoby żeby moja klasa `GpsApiLocationSource` rzucała wyjątkiem w przypadku rejecta? Twoja `IpBasedLocationSource` wydaje się już rzucać wyjątkiem przy reject, errory byłyby potem łapane przez try catche w metodach wyżej (`_getIPLocation` i `_getGpsLocation`).

Ewentualnie żeby nie zmieniać `GpsApiLocationSource`, można sprawdzać czy wynikiem `Promise.race` nie jest reject i dopiero w tym momencie rzucać error.

Reply:
Dzięki za uwagę, szczerze mi to umknęło, trochę nielogiczne żeby wyścig się kończył gdy ktoś odpadnie 😄 Dodałem nową metodę która działa jak `Promise.race` z tą różnicą, że nie przerywa działania gdy dostanie jakiegokolwiek rejecta. Nie używam jej jednak w przypadku GPSa, jako że tam `Promise.race` działa tak jak tego potrzebujemy.